File: python/examples/example-socket-subscriber.py
# SPDX-License-Identifier: BSL-1.0

import time
import traeger


def main():
    context: traeger.Context = traeger.Context()
    address: str = "tcp://localhost:5556"
    topics: list[str] = ["heart-beat"]

    subscriber: traeger.Subscriber = context.subscriber(address, topics)
    scheduler: traeger.Scheduler = traeger.Scheduler(threads_count=8)
    last_heart_beat = time.time()

    def on_heart_beat(topic, value):
        nonlocal last_heart_beat
        last_heart_beat = time.time()
        print(f"{topic}: {value}")

    promise = subscriber.listen(scheduler, on_heart_beat)
    print(f"Listening for heart-beat events on address: {address}")
    listening = True

    def stop(_):
        nonlocal listening
        listening = False

    promise.then_result(stop)
    while scheduler.count() != 0 and listening:
        try:
            now = time.time()
            if now - last_heart_beat >= 2:
                print("The last heart-beat was more than 2s ago")
                promise.set(traeger.Result.from_value(None))
            time.sleep(0.010)
        except KeyboardInterrupt:
            promise.set(traeger.Result.from_value(None))


if __name__ == "__main__":
    main()

File: rust/examples/example-socket-subscriber.rs
// SPDX-License-Identifier: BSL-1.0

use std::{
    cell::RefCell,
    rc::Rc,
    thread::sleep,
    time::{Duration, Instant},
};

use traeger::{Context, Scheduler, Threads};

fn main() {
    let context = Context::new();
    let address = "tcp://localhost:5556";
    let topics = ["heart-beat"];

    match context.subscriber(address, &topics) {
        Err(error) => {
            println!("Socket bind error: {error}");
        }
        Ok(subscriber) => {
            let scheduler = Scheduler::new(Threads { count: 8 });
            let last_heart_beat = Rc::new(RefCell::new(Instant::now()));

            println!("Listening for heart-beat events on address: {address}");
            let last_heart_beat_clone = last_heart_beat.clone();
            subscriber.listen(&scheduler, move |topic, variant| {
                *last_heart_beat_clone.borrow_mut() = Instant::now();
                println!("{topic}: {variant:?}");
            });

            while scheduler.count() != 0 {
                if last_heart_beat.borrow().elapsed().as_secs() >= 2 {
                    println!("The last heart-beat was more than 2s ago");
                    break;
                }
                sleep(Duration::from_millis(10));
            }
        }
    }
}
